stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/querychannels/internal/QueryChannelsStateLogicImpl.kt

```diff
@@ -0,0 +1,237 @@
+    }
+
+    /**
+     * Set the end of channels.
+     *
+     * @parami isEnd Boolean
+     */
+    override fun setEndOfChannels(isEnd: Boolean) {
+        mutableState._endOfChannels.value = isEnd
+    }
+
+    /**
+     * Sets if recovery is needed.
+     *
+     * @param recoveryNeeded Boolean
+     */
+    override fun setRecoveryNeeded(recoveryNeeded: Boolean) {
+        mutableState._recoveryNeeded.value = recoveryNeeded
+    }
+
+    /**
+     * Set the offset of the channels.
+     *
+     * @param offset Int
+     */
+    override fun setChannelsOffset(offset: Int) {
+        mutableState.channelsOffset.value = offset
+    }
+
+    /**
+     * Increments the channels offset.
+     *
+     * @param size Int
+     */
+    override fun incrementChannelsOffset(size: Int) {
+        val currentChannelsOffset = mutableState.channelsOffset.value
+        val newChannelsOffset = currentChannelsOffset + size
+        logger.v { "[updateOnlineChannels] newChannelsOffset: $newChannelsOffset <= $currentChannelsOffset" }
+        mutableState.channelsOffset.value = newChannelsOffset
+    }
+
+    /**
+     * MutableStateFlow<Boolean> for the current state. It returns the accordingly with loading first page
+     * or loading more channels.
+     */
+    override fun loadingForCurrentRequest(): MutableStateFlow<Boolean> {
+        return mutableState._currentRequest.value?.isFirstPage?.let { isFirstPage ->
+            if (isFirstPage) mutableState._loading else mutableState._loadingMore
+        } ?: mutableState._loading
+    }
+
+    /**
+     * Add channels to state
+     *
+     * @param channels List<Channel>.
+     */
+    override fun addChannelsState(channels: List<Channel>) {
+        mutableState.queryChannelsSpec.cids += channels.map { it.cid }
+        val existingChannels = mutableState.rawChannels ?: emptyMap()
+        mutableState.rawChannels = existingChannels + channels.map { it.cid to it }
+        channels.forEach { channel ->
+            logicRegistry.channelState(channel.type, channel.id).updateDataFromChannel(
+                channel,
+                shouldRefreshMessages = false,
+                scrollUpdate = false
+            )
+        }
+    }
+
+    /**
+     * Remove channels to state.
+     */
+    override fun removeChannels(cidSet: Set<String>) {
+        val existingChannels = mutableState.rawChannels ?: return
+
+        mutableState.queryChannelsSpec.cids = mutableState.queryChannelsSpec.cids - cidSet
+        mutableState.rawChannels = existingChannels - cidSet
+    }
+
+    /**
+     * Initializes [QueryChannelsMutableState.rawChannels] with an empty map if it wasn't initialized yet.
+     * This might happen when we don't have any channels in the offline storage and API request fails.
+     */
+    override fun initializeChannelsIfNeeded() {
+        if (mutableState.rawChannels == null) {
+            mutableState.rawChannels = emptyMap()
+        }
+    }
```

Comment:
Why don't we initialize it as an empty map directly?
Collections shouldn't have null values, they can always be an empty collection

Reply:
This changes the original behavior of the SDK. This method was already in QueryChannelsLogic. Let me change it and see if this doesn't break anything.